[Reland] [Inductor] Break the loop fusion when node2 depends on node1 mutations

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110677

Reland PR https://github.com/pytorch/pytorch/pull/109172 which has been reverted in https://github.com/pytorch/pytorch/pull/110622

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

Differential Revision: [D50097373](https://our.internmc.facebook.com/intern/diff/D50097373)

leslie-fang-intel committed:
[Reland] [Inductor] Break the loop fusion when node2 depends on node1 mutations

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @ckluk2, from the revert PR: https://github.com/pytorch/pytorch/pull/110622, you mentioned:

_Bisecting shows that this diff is the cause of S369683. Since this affects Ads production, need to back out this diff immediately.
Test Plan: See S369683_

May I know how can I check the detail failure message in ` S369683` and how to reproduce this failure?

ezyang commented:
It may not be so easy to get a repro (@ckluk2 you should see if you can trigger the minifier on the prod job, for this sort of thing it's typically pretty good), but here's the backtrace

```
  File "/dev/shm/uid-0/fbcbfadd-seed-nspid4026561868_cgpid26580064-ns-4026561865/torch/_inductor/compile_fx.py", line 340, in compile_fx_inner
    compiled_graph: CompiledFxGraph = fx_codegen_and_compile(
  File "/dev/shm/uid-0/fbcbfadd-seed-nspid4026561868_cgpid26580064-ns-4026561865/torch/_inductor/compile_fx.py", line 550, in fx_codegen_and_compile
    compiled_fn = graph.compile_to_fn()
  File "/dev/shm/uid-0/fbcbfadd-seed-nspid4026561868_cgpid26580064-ns-4026561865/torch/_inductor/graph.py", line 1001, in compile_to_fn
    return self.compile_to_module().call
  File "/dev/shm/uid-0/fbcbfadd-seed-nspid4026561868_cgpid26580064-ns-4026561865/torch/_dynamo/utils.py", line 190, in time_wrapper
    r = func(*args, **kwargs)
  File "/dev/shm/uid-0/fbcbfadd-seed-nspid4026561868_cgpid26580064-ns-4026561865/torch/_inductor/graph.py", line 952, in compile_to_module
    code, linemap = self.codegen()
  File "/dev/shm/uid-0/fbcbfadd-seed-nspid4026561868_cgpid26580064-ns-4026561865/torch/_inductor/graph.py", line 926, in codegen
    self.scheduler = Scheduler(self.buffers)
  File "/dev/shm/uid-0/fbcbfadd-seed-nspid4026561868_cgpid26580064-ns-4026561865/torch/_dynamo/utils.py", line 190, in time_wrapper
    r = func(*args, **kwargs)
  File "/dev/shm/uid-0/fbcbfadd-seed-nspid4026561868_cgpid26580064-ns-4026561865/torch/_inductor/scheduler.py", line 1109, in __init__
    self.fuse_nodes()
  File "/dev/shm/uid-0/fbcbfadd-seed-nspid4026561868_cgpid26580064-ns-4026561865/torch/_inductor/scheduler.py", line 1362, in fuse_nodes
    self.fuse_nodes_once()
  File "/dev/shm/uid-0/fbcbfadd-seed-nspid4026561868_cgpid26580064-ns-4026561865/torch/_inductor/scheduler.py", line 1375, in fuse_nodes_once
    for node1, node2 in self.get_possible_fusions():
  File "/dev/shm/uid-0/fbcbfadd-seed-nspid4026561868_cgpid26580064-ns-4026561865/torch/_inductor/scheduler.py", line 1424, in get_possible_fusions
    check_all_pairs(node_grouping)
  File "/dev/shm/uid-0/fbcbfadd-seed-nspid4026561868_cgpid26580064-ns-4026561865/torch/_inductor/scheduler.py", line 1411, in check_all_pairs
    if self.can_fuse(node1, node2):
  File "/dev/shm/uid-0/fbcbfadd-seed-nspid4026561868_cgpid26580064-ns-4026561865/torch/_inductor/scheduler.py", line 1529, in can_fuse
    if any(
  File "/dev/shm/uid-0/fbcbfadd-seed-nspid4026561868_cgpid26580064-ns-4026561865/torch/_inductor/scheduler.py", line 1532, in <genexpr>
    and node1.has_atomic_add(self.mutation_renames[node2_used_buf])
  File "/dev/shm/uid-0/fbcbfadd-seed-nspid4026561868_cgpid26580064-ns-4026561865/torch/_inductor/scheduler.py", line 704, in has_atomic_add
    return check_buf in self._get_atomic_add_buffers()
  File "/dev/shm/uid-0/fbcbfadd-seed-nspid4026561868_cgpid26580064-ns-4026561865/torch/_inductor/utils.py", line 335, in wrapper
    setattr(self, key, fn(self))
  File "/dev/shm/uid-0/fbcbfadd-seed-nspid4026561868_cgpid26580064-ns-4026561865/torch/_inductor/scheduler.py", line 687, in _get_atomic_add_buffers
    for node in self._body.get_nodes():
torch._dynamo.exc.BackendCompilerFailed: backend='compile_fn' raised:
AttributeError: 'NoneType' object has no attribute 'get_nodes'
```

This should probably be enough to backsolve a test case.

leslie-fang-intel committed:
Update on "[Reland] [Inductor] Break the loop fusion when node2 depends on node1 mutations"


Reland PR https://github.com/pytorch/pytorch/pull/109172 which has been reverted in https://github.com/pytorch/pytorch/pull/110622

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @ezyang, thanks for providing the backtrace. Looks like `self._body` could be `None` in this case. So, I add the check that `self._body` should be a instance of `ir.LoopBody`, before we use its `get_nodes()` method.
I think It should fix this issue. @ckluk2 Could you kindly help to take a try?

jansel commented:
@ckluk2 can you check that this fixes the issue and review?

ezyang commented on torch/_inductor/scheduler.py:
```diff
@@ -1519,6 +1551,25 @@ def can_fuse(self, node1: BaseSchedulerNode, node2: BaseSchedulerNode):
+        if (
+            isinstance(node1, (FusedSchedulerNode, SchedulerNode))
+            and isinstance(node2, SchedulerNode)
+            and isinstance(node2._body, ir.LoopBody)
```
@jansel do we really want to be doing pattern matches on scheduler nodes like this?

jansel replied:
There was some discussion on the prior PR.  I think including it in MemoryDeps is a bit cleaner... though both methods work.

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50097373).

ckluk2 commented:
I test this new version and the previous failure does not happen anymore.

leslie-fang-intel commented:
> I test this new version and the previous failure does not happen anymore.

Looks like it has fixed previous failure. Can I land this PR now? @ezyang

ezyang reviewed (approved):
ok

leslie-fang-intel commented:
@pytorchbot merge
